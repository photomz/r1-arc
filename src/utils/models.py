from src.dsl.arc_types import *
from typing import *
from dataclasses import dataclass
from types import CodeType
from functools import cached_property
from src.utils.types import deep_tuple
from src.utils.grid import FORMATTERS, FormatterNames
import inspect
from pydantic import BaseModel, Field


def grid_shape(grid: Grid) -> Shape:
    assert all(len(r) == len(r0 := grid[0]) for r in grid)
    return (len(grid), len(r0))


PuzzleSplit = Literal["train", "test"]  # ICL
DatasetSplit = Literal["train", "eval"]  # Dataset


@dataclass
class Example:
    I: Grid
    O: Grid
    split: PuzzleSplit
    n: int

    def __post_init__(self):
        # Assert I,O are tuples of tuples
        self.I = deep_tuple(self.I)
        self.O = deep_tuple(self.O)

        # assert I and O are Tuple[Tuple[int]], depth = 2
        # assert isinstance(self.I, tuple)
        # assert isinstance(self.O, tuple)
        # assert all(isinstance(r, tuple) for r in self.I)
        # assert all(isinstance(r, tuple) for r in self.O)
        # assert all(isinstance(c, int) for r in self.I for c in r)
        # assert all(isinstance(c, int) for r in self.O for c in r)

    @cached_property
    def I_shape(self):
        return grid_shape(self.I)

    @cached_property
    def O_shape(self):
        return grid_shape(self.O)

    def format(
        self,
        style=FormatterNames.SPREADSHEET,
        diff_style=FormatterNames.CELL_DIFF,
        no_output=False,
    ):
        fmt1 = FORMATTERS[style].format
        fmt2 = FORMATTERS[diff_style].diff

        s = f"Input {self.n+1}"
        s += "\n\n"
        s += fmt1(self.I, self.I_shape)

        if not no_output:
            s += "\n\n"
            s += f"Output {self.n+1}"
            s += "\n\n"
            s += fmt1(self.O, self.O_shape)

            if self.I_shape == self.O_shape:
                s += "\n\n"
                s += f"Diff {self.n+1} (I->O)"
                s += "\n\n"
                s += fmt2(self.I, self.O, self.I_shape, self.O_shape)

        return s

    def __repr__(self):
        return self.format()

    def __str__(self):
        return self.format()

    def dumps(self) -> Tuple[Grid, Grid]:
        return (deep_tuple(self.I), deep_tuple(self.O))


@dataclass
class TaskDef:
    id: str
    split: DatasetSplit
    train: List[Example]
    test: List[Example]

    solver: Optional[Callable] = None
    verifier: Optional[Callable] = None
    generator: Optional[Callable] = None

    @property
    def data(self):
        return self.train + self.test

    @property
    def solver_repr(self) -> str:
        # replace def solve_*(I): with def solve(I):
        try:
            source = inspect.getsource(self.solver)
        except TypeError:
            debug(f"Solver not found for {self.id}: {self.solver}")
            return None
        sanitized_source = source.replace(f"def solve_{self.id}(I):", "def solve(I):")
        return sanitized_source

    @property
    def verifier_repr(self) -> str:
        try:
            source = inspect.getsource(self.verifier)
        except TypeError:
            debug(f"Verifier not found for {self.id}: {self.verifier}")
            return None
        sanitized_source = source.replace(f"def verify_{self.id}", "def solve")
        return sanitized_source

    def format(
        self,
        hold_out=[],
        split: Literal["train", "test", "any"] = "any",
        no_output=False,
    ) -> str:
        arr = []
        if split == "train":
            arr = self.train
        elif split == "test":
            arr = self.test
        elif split == "any":
            arr = self.train + self.test
        s = ""
        for i, ex in enumerate(arr):
            # Positive and negative indexing
            if i in hold_out or i - len(arr) in hold_out:
                continue
            s += ex.format(no_output=no_output)
            s += "\n\n"
        return s

    def format_prompt(self):
        return (
            "## EXAMPLES\n"
            + self.format(split="train")
            + "\n## TEST\n"
            + self.format(split="test", no_output=True)
        )

    def dumps(self):
        return {
            "id": self.id,
            "train": [ex.dumps() for ex in self.train],
            "test": [ex.dumps() for ex in self.test],
            "solver": self.solver_repr if self.solver else None,
            "augment": None,
        }

    @classmethod
    def from_hf(self, x) -> "TaskDef":

        ex = {
            split: [
                Example(I=i, O=o, split=split, n=n) for n, (i, o) in enumerate(x[split])
            ]
            for split in ["train", "test"]
        }  # should recursively init, oh well depth = 1.

        return TaskDef(
            id=x["id"],
            split="split",
            train=ex["train"],
            test=ex["test"],
        )
