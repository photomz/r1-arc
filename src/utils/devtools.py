from enum import Enum
from typing_extensions import TypedDict
from devtools.debug import Debug, DebugArgument
from deepseek_tokenizer import ds_token
import tiktoken
from typing import Any, Optional, Callable
import typer


class TokenizerMethods(TypedDict):
    encode: Callable
    to_tokens: Callable


openai_coder = tiktoken.get_encoding("cl100k_base")

TOKENIZERS: dict[str, TokenizerMethods] = {
    "deepseek": {
        "encode": ds_token.encode,
        "to_tokens": ds_token.convert_ids_to_tokens,
    },
    "openai": {
        "encode": openai_coder.encode,
        "to_tokens": openai_coder.decode_single_token_bytes,
    },
}


class TokenArg(DebugArgument):
    def __init__(self, value: Any, *, name: Optional[str] = None, **extra: Any) -> None:
        super().__init__(value, name=name, **extra)
        if isinstance(value, str):
            n_tokens = len(TOKENIZERS["deepseek"]["encode"](value))
            self.extra.append(("tok", n_tokens))  # Deepseek tokens


class HotfixedDebug(Debug):
    output_class = Debug.output_class
    output_class.arg_class = TokenArg


debug = HotfixedDebug()


app = typer.Typer()


@app.command()
def tokenize(text: str, use: str = "deepseek"):
    """Count the number of tokens in a string"""
    coder = TOKENIZERS[use]
    ids = coder["encode"](text)
    tokens = list(map(coder["to_tokens"], ids))

    debug(tokens)


if __name__ == "__main__":
    app()
