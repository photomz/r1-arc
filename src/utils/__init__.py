from src.dsl.arc_types import Shape
from pathlib import Path
import orjson
import dataclasses
from src.utils.grid import FORMATTERS, FormatterNames

# root is recurse until .git found
ROOT = Path(__file__).resolve().parents[0]
while not (ROOT / ".git").exists():
    ROOT = ROOT.parent


def serialize(obj) -> str:
    if isinstance(obj, frozenset):
        return list(obj)
    elif isinstance(obj, set):
        return list(obj)
    elif isinstance(obj, Path):
        return str(obj)
    # elif is dataclass
    elif dataclasses.is_dataclass(obj):
        return dataclasses.asdict(obj)
    # Check if object is a pydantic model
    elif hasattr(obj, "dict") and callable(obj.dict):
        return obj.dict()
    elif callable(obj):
        return f"Function {obj.__name__}"

    raise TypeError(f"Object of type {type(obj)} is not JSON serializable: {obj}")


def jdumps(data: dict) -> str:
    return orjson.dumps(data, option=orjson.OPT_INDENT_2, default=serialize).decode()


def get_shape(O) -> Shape:
    n = len(O)
    m = len(O[0])
    assert all(len(row) == m for row in O)
    return (n, m)
