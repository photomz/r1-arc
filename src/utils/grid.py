from abc import ABC, abstractmethod
from dataclasses import dataclass
from collections import defaultdict
from enum import Enum
from typing import TypeVar
from itertools import product
import numpy as np
import typer
from pathlib import Path
import json
from devtools import debug

from src.utils.devtools import debug
from src.dsl.arc_types import *

# (A - Z) + (AA - AD)
ALPHA = list(chr(ord("A") + i) for i in range(26)) + list(
    f"A{chr(ord('A') + i)}" for i in range(4)
)


def bucket_diffs(g1: Grid, g2: Grid) -> dict[tuple[int, int], list[tuple[int, int]]]:
    """Bucket diffs by index (color)"""
    x, y = (g1 != g2).nonzero()
    diffs = defaultdict(list)

    for i, j in zip(x, y, strict=True):
        pair = (g1[i][j], g2[i][j])
        diffs[pair].append((int(i), int(j)))

    return diffs


@dataclass
class Formatter(ABC):
    sep: str = "|"
    newline: str = "\n"
    n_max: int = 30

    @classmethod
    def format(cls, grid: Grid, shape: Shape) -> str:
        """Base formatter"""
        if shape[0] > cls.n_max or shape[1] > cls.n_max:
            raise ValueError(f"Grid dimensions exceed maximum size of {cls.n_max}")
        return ""


@dataclass
class GridFormatter(Formatter):
    """
    0 | 0 | 0
    0 | 1 | 0
    0 | 0 | 0
    """

    @classmethod
    def format(cls, grid: Grid, shape) -> str:
        super().format(grid, shape)
        return cls.newline.join(cls.sep.join(str(x) for x in row) for row in grid)


@dataclass
class SpreadsheetFormatter(Formatter):
    """
        A | B | C
    1	0 | 0 | 0
    2	0 | 1 | 0
    3 	0 | 0 | 0
    """

    @classmethod
    def format(cls, grid: Grid, shape) -> str:
        super().format(grid, shape)
        _, n_cols = shape

        header = "\t" + cls.sep.join(ALPHA[:n_cols])
        rows = "\n".join(
            str(i + 1) + "\t" + cls.sep.join([str(x) for x in row])
            for i, row in enumerate(grid)
        )
        return cls.newline.join([header, rows])


@dataclass
class Differ(Formatter):
    change: str = "->"
    blank: str = "    "

    @classmethod
    def diff(self, A: Grid, B: Grid, A_shape, B_shape) -> str:
        """Base diff 2 grids"""
        super().format(A, A_shape)
        super().format(B, B_shape)

        if A_shape != B_shape:
            raise ValueError(f"Unexpected shape: A {A_shape}, B {B_shape}")
        return ""


@dataclass
class ListDiffer(Differ):

    @classmethod
    def diff(self, A: Grid, B: Grid, a, b) -> str:
        """List A->B color diffs like adjacency list"""
        super().diff(A, B, a, b)

        diffs = bucket_diffs(A, B)
        output = []

        for (cA, cB), coords in sorted(diffs.items()):
            cells = " ".join(f"{ALPHA[i]}{j+1}" for i, j in sorted(coords))
            output.append(f"{cA}{self.change}{cB}: {cells}")

        return "\n".join(output)


@dataclass
class CellDiffer(Differ):

    @classmethod
    def diff(self, A: Grid, B: Grid, A_shape, B_shape) -> str:
        """Normal grid, but each cell is (ij)->(ij)' diff"""
        super().diff(A, B, A_shape, B_shape)

        rows, cols = A_shape
        # Init with blank cells
        diff_grid = [[self.blank for _ in range(cols)] for _ in range(rows)]

        for i, j in product(range(rows), range(cols)):
            if A[i][j] != B[i][j]:
                diff_grid[i][j] = f"{A[i][j]}{self.change}{B[i][j]}"

        make_row = lambda row: self.sep.join(str(x) for x in row)

        return self.newline.join(map(make_row, diff_grid))


class FormatterNames(str, Enum):
    GRID = "grid"
    SPREADSHEET = "spreadsheet"
    LIST_DIFF = "diff-list"
    CELL_DIFF = "diff-cell"


TASK_PATH = Path(__file__).parent.parent.parent / "tasks"
FORMATTERS = {
    FormatterNames.GRID: GridFormatter,
    FormatterNames.SPREADSHEET: SpreadsheetFormatter,
    FormatterNames.LIST_DIFF: ListDiffer,
    FormatterNames.CELL_DIFF: CellDiffer,
}
