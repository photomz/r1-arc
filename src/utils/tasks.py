from src.dsl.arc_types import *
from typing import *
import src.dsl.solvers as solvers_module
import src.re_arc.verifiers as verifiers_module
import src.re_arc.generators as generators_module

from src.utils.types import ROOT
import orjson as json
import inspect
from src.utils.models import Example, TaskDef
from src.utils.types import deep_tuple

DATA_DIR = ROOT / "src/data"


def get_tasks() -> Dict[str, TaskDef]:
    tasks = {}
    for split in {"train", "eval"}:
        with (DATA_DIR / f"{split}.jsonl").open() as f:
            lines = [json.loads(line) for line in f]
        for l in lines:
            id = l["task_id"]
            solver, verifier, generator = None, None, None
            if split == "train":
                solver = getattr(solvers_module, f"solve_{id}", None)
                verifier = getattr(verifiers_module, f"verify_{id}", None)
                generator = getattr(generators_module, f"generate_{id}", None)
            tasks[id] = TaskDef(
                id=id,
                split=split,
                train=[
                    Example(
                        I=deep_tuple(ex["input"]),
                        O=deep_tuple(ex["output"]),
                        split="train",
                        n=i,
                    )
                    for i, ex in enumerate(l["train"])
                ],
                test=[
                    Example(
                        I=deep_tuple(ex["input"]),
                        O=deep_tuple(ex["output"]),
                        split="test",
                        n=i,
                    )
                    for i, ex in enumerate(l["test"])
                ],
                solver=solver,
                verifier=verifier,
                generator=generator,
            )

    return tasks


TASKS = get_tasks()

if __name__ == "__main__":
    debug(list(TASKS.items())[:4])
