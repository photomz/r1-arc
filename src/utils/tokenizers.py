from textual.app import App, ComposeResult
from textual.containers import Container
from textual.widgets import Header, Input, Pretty
from textual.reactive import reactive
from typing import List

from src.utils.devtools import TOKENIZERS


class TokenCounterApp(App):
    """A Textual app that shows token counts and tokens in real-time."""

    # Reactive variable to store the current input text
    current_text = reactive("")

    def compose(self) -> ComposeResult:
        yield Header()
        yield Container(
            Input(placeholder="Type text to analyze tokens...", id="text_input"),
            Pretty([]),  # Will hold our token analysis
            id="main_container",
        )

    def on_input_changed(self, event: Input.Changed) -> None:
        """Update the token analysis whenever the input changes."""
        self.current_text = event.value

    def watch_current_text(self, text: str) -> None:
        """React to changes in the input text."""
        analysis = []

        for tokenizer_name, tokenizer in TOKENIZERS.items():
            tokens = tokenizer["encode"](text)
            token_strings = [tokenizer["to_tokens"](token) for token in tokens]

            result = {
                "tokenizer": tokenizer_name,
                "token_count": len(tokens),
                "tokens": token_strings,
            }
            analysis.append(result)

        # Update the Pretty widget with the new analysis
        pretty_widget = self.query_one(Pretty)
        pretty_widget.update(analysis)


if __name__ == "__main__":
    app = TokenCounterApp()
    app.run()
