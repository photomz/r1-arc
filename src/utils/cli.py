from abc import ABC, abstractmethod
from dataclasses import dataclass
from collections import defaultdict
from enum import Enum
from typing import TypeVar
from itertools import product
import numpy as np
import typer
from pathlib import Path
import json
from devtools import debug

from src.utils.tasks import TASKS
from src.utils.devtools import debug
from src.dsl.arc_types import *
from src.utils.grid import *
import random

app = typer.Typer()
random_id = int(random.random() * 800)


@app.command()
def format(id: str = None, style: FormatterNames = FormatterNames.GRID):
    """Format or diff all grids in ARC task"""
    formatter = FORMATTERS[style]
    is_diff = style in {FormatterNames.LIST_DIFF, FormatterNames.CELL_DIFF}

    if id is None:
        id = list(TASKS.values())[random_id].id
    grids = TASKS[id].train + TASKS[id].test

    for i, g in enumerate(grids, 1):
        print(f"> Grid {i if i < len(grids) else 'TEST'}")

        A = np.array(g.I)
        B = np.array(g.O)

        # B.size skips held-out test grid output
        if is_diff:
            diff = formatter.diff(A, B, A.shape, B.shape)
            debug(diff)
        else:
            finput = formatter.format(A, A.shape)
            debug(finput)

            foutput = formatter.format(B, B.shape)
            debug(foutput)
    debug(id)

    print(TASKS[id].format_prompt())


if __name__ == "__main__":
    app()
