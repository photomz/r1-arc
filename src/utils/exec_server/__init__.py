"""
Python Execution Server for ARC

This module provides a secure environment for executing Python code in ARC problems.
It includes:
- Code execution with resource limits
- Static analysis of code
- Error detection and auto-repair
- Caching for performance
- CLI for running solvers on tasks
"""

# Import main components for easy access
from src.utils.exec_server.execution import (
    run_python_transform_sync,
    run_python_transform_async,
    run_python_transforms,
    PythonResult,
    PythonException,
    ResourceLimits,
    TransformInput
)

from src.utils.exec_server.code_analysis import (
    CodeMetrics,
    collect_static_metrics,
    attempt_code_repair
)

# Import CLI
from src.utils.exec_server.cli import app as cli_app

# Version info
__version__ = "0.1.0"