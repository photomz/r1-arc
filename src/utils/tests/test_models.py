from src.utils import TASKS


def test_many_test_examples(snapshot):
    """Should not leak test outputs"""
    assert TASKS["794b24be"].format_prompt() == snapshot
