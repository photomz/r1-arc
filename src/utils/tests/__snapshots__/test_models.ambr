# serializer version: 1
# name: test_many_test_examples
  '''
  ## EXAMPLES
  Input 1
  
  	|A|B|C
  1	0|0|0
  2	1|0|0
  3	0|0|0
  
  Output 1
  
  	|A|B|C
  1	2|0|0
  2	0|0|0
  3	0|0|0
  
  Diff 1 (I->O)
  
  0 -> 2|  --  |  --  
  1 -> 0|  --  |  --  
    --  |  --  |  --  
  
  Input 2
  
  	|A|B|C
  1	0|1|0
  2	1|0|0
  3	0|0|0
  
  Output 2
  
  	|A|B|C
  1	2|2|0
  2	0|0|0
  3	0|0|0
  
  Diff 2 (I->O)
  
  0 -> 2|1 -> 2|  --  
  1 -> 0|  --  |  --  
    --  |  --  |  --  
  
  Input 3
  
  	|A|B|C
  1	0|0|1
  2	0|0|0
  3	1|0|0
  
  Output 3
  
  	|A|B|C
  1	2|2|0
  2	0|0|0
  3	0|0|0
  
  Diff 3 (I->O)
  
  0 -> 2|0 -> 2|1 -> 0
    --  |  --  |  --  
  1 -> 0|  --  |  --  
  
  Input 4
  
  	|A|B|C
  1	0|1|0
  2	0|0|1
  3	0|0|0
  
  Output 4
  
  	|A|B|C
  1	2|2|0
  2	0|0|0
  3	0|0|0
  
  Diff 4 (I->O)
  
  0 -> 2|1 -> 2|  --  
    --  |  --  |1 -> 0
    --  |  --  |  --  
  
  Input 5
  
  	|A|B|C
  1	0|0|1
  2	0|0|0
  3	0|0|0
  
  Output 5
  
  	|A|B|C
  1	2|0|0
  2	0|0|0
  3	0|0|0
  
  Diff 5 (I->O)
  
  0 -> 2|  --  |1 -> 0
    --  |  --  |  --  
    --  |  --  |  --  
  
  Input 6
  
  	|A|B|C
  1	1|1|0
  2	0|0|0
  3	1|0|0
  
  Output 6
  
  	|A|B|C
  1	2|2|2
  2	0|0|0
  3	0|0|0
  
  Diff 6 (I->O)
  
  1 -> 2|1 -> 2|0 -> 2
    --  |  --  |  --  
  1 -> 0|  --  |  --  
  
  Input 7
  
  	|A|B|C
  1	0|1|0
  2	1|1|0
  3	0|0|0
  
  Output 7
  
  	|A|B|C
  1	2|2|2
  2	0|0|0
  3	0|0|0
  
  Diff 7 (I->O)
  
  0 -> 2|1 -> 2|0 -> 2
  1 -> 0|1 -> 0|  --  
    --  |  --  |  --  
  
  Input 8
  
  	|A|B|C
  1	1|1|0
  2	0|0|0
  3	1|0|1
  
  Output 8
  
  	|A|B|C
  1	2|2|2
  2	0|2|0
  3	0|0|0
  
  Diff 8 (I->O)
  
  1 -> 2|1 -> 2|0 -> 2
    --  |0 -> 2|  --  
  1 -> 0|  --  |1 -> 0
  
  Input 9
  
  	|A|B|C
  1	0|1|0
  2	1|1|0
  3	1|0|0
  
  Output 9
  
  	|A|B|C
  1	2|2|2
  2	0|2|0
  3	0|0|0
  
  Diff 9 (I->O)
  
  0 -> 2|1 -> 2|0 -> 2
  1 -> 0|1 -> 2|  --  
  1 -> 0|  --  |  --  
  
  Input 10
  
  	|A|B|C
  1	1|0|0
  2	0|0|1
  3	0|1|1
  
  Output 10
  
  	|A|B|C
  1	2|2|2
  2	0|2|0
  3	0|0|0
  
  Diff 10 (I->O)
  
  1 -> 2|0 -> 2|0 -> 2
    --  |0 -> 2|1 -> 0
    --  |1 -> 0|1 -> 0
  
  
  ## TEST
  Input 1
  
  	|A|B|C
  1	0|1|0
  2	0|0|0
  3	0|1|0
  
  Input 2
  
  	|A|B|C
  1	0|1|0
  2	0|1|1
  3	1|0|0
  
  
  '''
# ---
